### ISSUE TITLE
listener.ws.external.proxy_protocol can not work, trigger error:[error] Ranch listener 'mqtt:ws:8083' had connection

### ISSUE THREAD

**Issue by bibitiger** [emqx/emqx#4616]
**Please describe your problem in detail, if necessary, you can upload the log file through the attachment**:
os:
```
NAME="Ubuntu"
VERSION="16.04.7 LTS (Xenial Xerus)"
```
emqx: `4.2.7`
nginx:`openresty/1.7.7.1 core:nginx 1.7.7`

error msg:
```
2021-04-17 20:52:45.617 [error] Ranch listener 'mqtt:ws:8083' had connection process started with cowboy_clear:start_link/4 at <0.1763.0> exit with reason: {{badmatch,{error,protocol_error,'The PROXY protocol header signature was not recognized. (PP 2.1, PP 2.2)'}},[{cowboy_clear,connection_process,4,[{file,"cowboy_clear.erl"},{line,38}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
```

emqx.conf
```
## The access control for the MQTT/WebSocket listener.
##
## See: listener.ws.$name.access
##
## Value: ACL Rule
listener.ws.external.access.1 = allow all

## Verify if the protocol header is valid. Turn off for WeChat MiniApp.
##
## Value: on | off
listener.ws.external.verify_protocol_header = off

## Enable the Proxy Protocol V1/2 if the EMQ cluster is deployed behind
## HAProxy or Nginx.
##
## See: listener.ws.$name.proxy_protocol
##
## Value: on | off
listener.ws.external.proxy_protocol = on
```

nginx.conf
```
server {
        listen 443 ssl proxy_protocol;

        ssl on;
        server_name wss8084.megahealth.cn;

        ssl_certificate   /home/lin/work/cert/4014180__megahealth.cn.pem;
        ssl_certificate_key  /home/lin/work/cert/4014180__megahealth.cn.key;
        ssl_session_timeout 5m;
        ssl_ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE:ECDH:AES:HIGH:!NULL:!aNULL:!MD5:!ADH:!RC4;
        ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
        ssl_prefer_server_ciphers on;
        error_page 497  https://$host$uri?$args;

        location /mqtt {
            proxy_redirect off;
            proxy_pass http://127.0.0.1:8083; 
            proxy_set_header Host $host;
            proxy_set_header X-Real_IP $remote_addr;
            proxy_set_header X-Forwarded-For $remote_addr:$remote_port;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection upgrade;
        }
    }
```

please help me. I need devices true address.

**Comment by sh3d2** [comment]
Hi,

dont confuse custom headers (x-forwarded-for/x-real-ip etc) with proxy protocol https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt (proxy-protocol works at the tcp level) and AFAIK nginx doesnt support writing proxy protocol v2 info into upstream (in your case that is emqx server) https://trac.nginx.org/nginx/ticket/1639. and emqx dropped custom headers support with 4.0 version 
https://github.com/emqx/emqx/issues/1335
https://github.com/emqx/emqx/issues/3003 

I'd recommend using haproxy for passing proxy-protocol info https://serverfault.com/questions/839965/send-proxy-protocol-header-from-haproxy

TL;DR use haproxy for proper proxy-protocol load balancing

**Comment by bibitiger** [comment]
@sh3d2 Hi,
some online service is already on the server by nginx.switch to haproxy is  risky.is there other way to complete?

**Comment by sh3d2** [comment]
@bibitiger emqx 4.3 introduces http proxy via headers https://github.com/emqx/emqx/issues/4184 

that way setting 

```
listener.ws.external.proxy_address_header = x-real-ip // this one you have in your nginx config
```
will give you client real ip address.

so your options are:
* upgrade emqx to 4.3 and use headers proxy
* drop nginx and go with haproxy to use proxy protocol

**Comment by bibitiger** [comment]
@sh3d2 4.3 is useful, wonderful. thanks very much!

**Comment by zmstone** [comment]
Thank you for your support @sh3d2

**Comment by diabl0w** [comment]
Can someone please help me out? I have the same problem of proxy_protocol issues creating error:

` 2021-04-17 20:52:45.617 [error] Ranch listener 'mqtt:ws:8083' had connection process started with cowboy_clear:start_link/4 at <0.1763.0> exit with reason: {{badmatch,{error,protocol_error,'The PROXY protocol header signature was not recognized. (PP 2.1, PP 2.2)'}},[{cowboy_clear,connection_process,4,[{file,"cowboy_clear.erl"},{line,38}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}`

I need to keep nginx, but using the proxy address headers fix is not working. I am using the latest docker image which runs 4.3.1. ngnix config:

`server {
        listen 8083;
        server_name *dns name*;

        location / {
            proxy_redirect off;
            proxy_pass http://192.168.1.101:8083/; 
            proxy_set_header Host $host;
            proxy_set_header X-Real_IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection upgrade;
        }
    }`

For emqx I have tried both
`listener.ws.external.proxy_address_header = X-Real_IP`
and
`listener.ws.external.proxy_address_header = X-Forwarded-For`

**Comment by zmstone** [comment]
Hi @diabl0w 
it seems the server is trying to read the haproxy binary protocol.
have you set this config?
`listener.ws.external.proxy_protocol = on`

Since you are using http headers for source info, you can turn off the tcp layer binary protocol.

**Comment by diabl0w** [comment]
@zmstone thank you for the prompt reply. yes, I do have `listener.ws.external.proxy_protocol = on` set. To explain with more details.. my current relevant emqx config:
```
listener.ws.external = 8083
listener.ws.external.mqtt_path = /
listener.ws.external.zone = external
listener.ws.external.access.1 = allow all
listener.ws.external.proxy_protocol = on
listener.ws.external.proxy_protocol_timeout = 5s
```
relevant nginx config:
```
server {
    listen       8083 proxy_protocol;
    server_name  s0m1.farted.net;

    location / {
        proxy_pass http://192.168.111.101:8083/;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
        proxy_set_header Host $host;
        proxy_set_header X-Real_IP $remote_addr;
        proxy_pass_request_headers on;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
```
The exact error I receive in emqx debug log:
```

2021-05-23T04:04:53.498303+00:00 [error] crasher: initial call: cowboy_clear:connection_process/4, pid: <0.2002.0>, registered_name: [], exit: {{recv_proxy_header,'The PROXY protocol header signature was not recognized. (PP 2.1, PP 2.2)'},[{cowboy_clear,connection_process,4,[{file,"cowboy_clear.erl"},{line,41}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}, ancestors: [<0.1854.0>,<0.1853.0>,ranch_sup,<0.1557.0>], message_queue_len: 1, messages: [{handshake,'mqtt:ws:8083',ranch_tcp,#Port<0.23>,5000}], links: [<0.1854.0>,#Port<0.23>], dictionary: [], trap_exit: false, status: running, heap_size: 610, stack_size: 28, reductions: 301; neighbours:


2021-05-23T04:04:53.498626+00:00 [error] Ranch listener 'mqtt:ws:8083' had connection process started with cowboy_clear:start_link/4 at <0.2002.0> exit with reason: {recv_proxy_header,'The PROXY protocol header signature was not recognized. (PP 2.1, PP 2.2)'}
```

I am not sure what "you can turn off the tcp layer binary protocol" means, unfortunately I am sorry. 
As i said, I am also open to using the headers solution, as long as I can get external websockets working I don't care which way it works.

**Comment by zmstone** [comment]
I mean you should set `off` for that config

**Comment by fernando-almeida** [comment]
I'm currently hitting this issue and can't connect using websockets to EMQx.
Would it possible to provide a minimal working configuration to make this work for a K8s setup that's use Nginx as proxy?
I've carefully read through all the details of this issue and also #4184 and I am unable to get this to work even after some experimentation with the targetted variables.
In a nutshell without  `listener.ws.external.proxy_protocol = on` I don't even get any kind logs from the server even with log.level set to `debug`.   
Thanks in advance for any help.

**Comment by zmstone** [comment]
hi @fernando-almeida 
Which version are you using?

Websocket proxy can work on two network layers: Http, and tcp
Emqx supports proxy protocol v2 (the binary protocol at tcp layer) with the listener.ws.external.proxy_protocol config.

Otherwise if it’s at http layer, http headers are used to recognize the source info. E.g X-Forwarded-For 

Please, could you be more specific on which approach makes more sense to you?

Here is an article for nginx proxy at http layer.
https://www.emqx.com/en/blog/connect-to-mqtt-broker-with-websocket

here is one for working at tcp layer 
https://emqx.medium.com/emq-x-message-server-nginx-reverse-proxy-8e5883bf357d
In this case websocket is transparent to nginx.

**Comment by fernando-almeida** [comment]
hi @zmstone.
Thanks for the quick-reply.
Some additional context about my setup.
I'm using a K8s cluster (v1.21.5) on Digital Ocean (DO).
I'm trying to rely on the [emqx/emqx-rel Helm chart ](https://github.com/emqx/emqx-rel).
I'm currently relying on [nginx-ingress](https://kubernetes.github.io/ingress-nginx/) Helm chart so for the time being I want to try to stick with nginx.

It looks like I have inadvertently messed up the K8s Load Balancer service configuration and the nodeport were not matching the one I had explicitly set up on the DO Load Balancer.

I'm now able to connect externally using EMQx via all protocols, just not wss which I'm obliged to use I believe when service an SAP from an HTTPs endpoint.
When trying to connect using Mqtt.js CLI with the websocket protocol with security enabled I'm getting this error:

not an SSL/TLS record: 485454502f312e31203430302042616420526571756573740d0a636f6e6e656374696f6e3a20636c6f73650d0a636f6e74656e742d6c656e6774683a20300d0a0d0a

For context, I was/am assuming that TLS termination could/can take place at the DO load balancer.
So as far as port forwarding goes this is what I have set up for the standard WSS port:
- DO Load Balancer Forwarding rules (8084 -> 30184, 8083 -> 30183)
- K8S Nginx Load Balancer Service port mapping:
```
ports:
 - name: 8083-tcp
      protocol: TCP
      port: 8083
      targetPort: 8083-tcp
      nodePort: 30183 
 - name: 8084-tcp
      protocol: TCP
      port: 8084
      targetPort: 8084-tcp
      nodePort: 30184 
```
- Nginx TCP service config port mapping (bound to nginx-ingress-controller):
```
8083: dev/emqx-headless:8083
8084: dev/emqx-headless:8083
```

I was expecting that I would be able to set it up like this and still be able to use the secure Websocket protocol. 
The links you provided at least the way I perceive them don't explicitly address this specific use case. For additional context the external TLS connection is established from a Let's Encrypt certificate that's being set on K8s on a per-ingress basis depending on the public domain I'm currently using.

Thanks again for your quick reply and concise guidance

**Comment by zmstone** [comment]
emqx-rel is deprecated.
If you are on emqx 4.3, please find the latest chat here https://github.com/emqx/emqx/tree/main-v4.3/deploy/charts/emqx

“not a tls record” error message hints a tls handshake error, maybe a misconfiguration in where TLS is terminated (DO LB).

> 8084 -> 30184 , 8083 -> 30183)

if you’d *terminate* TLS at LB, the LB it should communicate with the backend over plaintext TCP, there should *no* need for an ssl listener in EMQ X.

**Comment by fernando-almeida** [comment]
Hi @zmstone 
Both places reference the same chart repo (https://repos.emqx.io/charts) and thus I believe I'm already using the latest chart/version.
For reference I arrived at [emqx/emqx-chart](https://github.com/emqx/emqx-chart) and then was pointed to [emqx/emq-rel](https://github.com/emqx/emqx-rel/tree/master/deploy/charts/emqx) and now you're pointing me to [here](https://github.com/emqx/emqx/blob/main-v4.3/deploy/charts/emqx/templates/StatefulSet.yaml). Maybe it would be best to clean up / downright remove the aforementioned repos to avoid confusion in the future. There aren't that much significant differences from what's on the link to repo you provided apart for some additional volume mounts from configMaps and some typos being fixed.

I'm inclined to agree with you regarding the last statement but configuration wise I'm unsure what I should be putting in the ingress to allow the /mqtt path to be propertly forwarded to the EMQx broker. Ideally, if I'd expose port 8084 to the outside world on the LB and just mapping to the internal EMQx service that listens for ws that should suffice.
I'm going to review my ingress settings and try to work around this as I currently have a separate ingress that I'd would expect to be picked for bridging to the internal service. Its configuration is the following
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: emqx-websocket
  annotations:
    nginx.ingress.kubernetes.io/configuration-snippet: |
      # proxy_protocol on;
      proxy_set_header Host $host;
      # Reverse proxy retains client address
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-Port $server_port;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      # proxy_set_header X-Forwarded-Proto $scheme;
      # WebSocket extra request header
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
    # nginx.ingress.kubernetes.io/proxy-redirect-from: "off"
    # nginx.ingress.kubernetes.io/proxy-redirect-to: "off"
    nginx.ingress.kubernetes.io/connection-proxy-header: keep-alive
    nginx.ingress.kubernetes.io/proxy-connect-timeout: '120'
    nginx.ingress.kubernetes.io/proxy-http-version: "1.1"
    nginx.ingress.kubernetes.io/proxy-read-timeout: '3600'
    nginx.ingress.kubernetes.io/proxy-send-timeout: '3600'
spec:
  tls:
    - hosts:
        - xxx.example.com
      secretName: wilcard-xxx-example-com-tls
  rules:
    - host: xxx.example.com
      http:
        paths:
          - path: /mqtt
            pathType: ImplementationSpecific
            backend:
              service:
                name: sense-emqx
                port:
                  number: 8083
```

Thanks again for the insights

**Comment by zmstone** [comment]
> Maybe it would be best to clean up / downright remove the aforementioned repos to avoid confusion in the future.

Thank you for your suggestion, we'll update READMEs.

`Client -- DO LB -- Ingress -- EMQ X`
If `DO LB` terminates TLS, everything behind should be plaintext,
not sure why you still need a tls spec for ingress.

**Comment by fernando-almeida** [comment]
Hi @zmstone,

I still need to use SSL/TLS in some capacity externally, no?
My client app (VueJS SPA) is served via HTTPs which implies I'm required to use wss:// and not ws:// to connect to the websocket via the external endpoint.
Having the ingress set up the way I described was the only I found in order to that along with forwarding relevant HTTP custom headers so I can see the real client IP/port data on EMQx.
The lines in the configuration snippet preceded with the pound sign (commented) were a desperate experience to try to make this work. I'm unaware if websockets protocol requires any further specific data being forwarded from the LB to the backend service that listens for websocket connections.

I mostly got these ideas from the [configuration docs](https://www.emqx.io/docs/en/v4.3/configuration/configuration.html#authacl)
```
listener.ws.external.proxy_address_header
listener.ws.external.proxy_port_header
```

Some other ideas I got came from [here](https://titanwolf.org/Network/Articles/Article?AID=8772c6a8-6361-4138-aad7-a7ef7cce2cfc) and the issues in the repo #1335, #3003 and also #4616

**Comment by zmstone** [comment]
> I still need to use SSL/TLS in some capacity externally, no?

Not if all your traffic are routed through digital ocean LB which terminates TLS.

Also assuming DigitalOcean LB works at transport layer.
Assuming this setup: `Client --(wss)-- DigitalOcean LB --(TCP with PPv2 tcp headers)-- Ingress(Nginx) --(ws with http headers)-- EMQ X`

I' not entirely sure how `DO LB` works, but if it send to `Ingress` TCP with proxy protocol v2 header,
then ingress should be configured to accept proxy protocol.
I found this article https://docs.nginx.com/nginx/admin-guide/load-balancer/using-proxy-protocol/

**Comment by fernando-almeida** [comment]
Hi @zmstone,

Sorry, my bad.
My actual setup is
Also assuming this setup:
Client --(wss)-- DigitalOcean LB --(L4/L7 Bridge)-- Ingress-Controller (nginx-ingress community) --(ws with http headers)-- EMQ X

The SSL termination does not take place at DO LB but it takes place at the nginx-ingress controller at least for HTTPS. This is by design as I'm not using DO DNS to automatically manage Let's Encrypt cert renewals (required for this) but rather I'm manually setting these up or relying on cert-manager for this purpose depending on the domain.
I don't think one can setup up TCP SSL termination on a per-service basis using the community nginx-ingress, to the best of my knowledge you can only route the requests to the backend via a [config map](https://kubernetes.github.io/ingress-nginx/user-guide/exposing-tcp-udp-services/).
DO supports enabling proxy protocol but that just messes up all the remainder HTTP/HTTPS traffic I need to get to API servers I'm running in the cluster. I was expecting that the use of the [use-proxy-protocol](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/configmap/#use-proxy-protocol) annotation for the ingress would help with this but it does not seem to be the case. My goal was to try and enable the proxy protocol on a per-ingress basis because it supposedly defaults to **false** but that did not work as intended.
Ultimately nginx [supports it though](https://docs.nginx.com/nginx/admin-guide/security-controls/terminating-ssl-tcp/).

Will keep trying and get back to you here if I succeed.
Once again I appreciate all of your insights.

**Comment by fernando-almeida** [comment]
Hi @zmstone,

My issue has been sorted out.
I ended up messing several configs inadvertently that I shouldn't have which I describe below so anyone can avoid these and/or rule them out when debugging:
- Using the incorrect port for the K8s EMQx service that mapped to the external WS listener as configured in the container.
- "Bug" with MQTT.js CLI that seems incapable in MacOS 10.15.7 to properly validate the certificate chain retrieved by the based on a Let's Encrypt issued certificate unless we provide it with an explicit path and/or file to the CAs to check.
```
Client 'UNKNOWN@myhost.example.com' DISCONNECTED unable to find valid certification path to requested target
PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
```

The specific details about my K8s setup at the time I got this to work are:
- K8s Server GitVersion:"v1.21.5", GitCommit:"aea7bbadd2fc0cd689de94a54e5b7b758869d691
- nginx-ingress Helm Chart (community version ingress-nginx-3.40.0, AppVersion=0.50.0, controllerImage=
k8s.gcr.io/ingress-nginx/controller:v0.47.0@sha256:a1e4efc107be0bb78f32eaec37bef17d7a0c81bec8066cdf2572508d21351d0b)
- emqx-4.4.0 Helm Chart (modified to make use of the emqx/emqx:4.3.11 docker image)

The ultimate configuration that worked for me is the following in case you're interested in adding it to the docs
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: emqx-wss
  namespace: dev
  labels:
    app.kubernetes.io/instance: emqx
    app.kubernetes.io/name: emqx
  annotations:
    meta.helm.sh/release-name: emqx
    meta.helm.sh/release-namespace: dev
    nginx.ingress.kubernetes.io/backend-protocol: "http"
    nginx.ingress.kubernetes.io/use-forwarded-headers: "true"
    nginx.ingress.kubernetes.io/enable-real-ip: "true"
    nginx.ingress.kubernetes.io/proxy-request-buffering: "off"
    nginx.ingress.kubernetes.io/proxy-connect-timeout: "120"
    nginx.ingress.kubernetes.io/proxy-http-version: "1.1"
    nginx.ingress.kubernetes.io/proxy-read-timeout: "3600"
    nginx.ingress.kubernetes.io/proxy-send-timeout: "3600"
    nginx.ingress.kubernetes.io/use-proxy-protocol: "false"
    nginx.ingress.kubernetes.io/proxy-protocol-header-timeout: "5s"
spec:
  tls:
    - hosts:
        - myhost.example.com # Change to your specific host name
      secretName: my-host-tls # Secret that holds the K8s certificates for the host (confirmed to work with both single and wildcard certs)
  rules:
    - host: myhost.example.com
      http:
        paths:
          - path: /mqtt # Default in EMQx but make sure to match it if it changes
            pathType: Prefix
            backend:
              service:
                name: emqx
                port:
                  number: 8083 # EMQx WS external listening port
```

I've forked the [emqx repo](https://github.com/fernando-almeida/emqx) and included some additional changes to the chart based off the master branch that I believe could bring some additional flexibility that may be needed. I'm happy to incorporate a config setting in the Chart to allow users to more seamlessly have integrated with websocket with SSL termination at the nginx-ingress-controller.
Let me know if that's desirable I'm happy to make a PR and follow any additional guidance you may have.
